### PR TITLE
[Collections/split] fix `.path` behavior (including bug on Windows)

### LINKED ISSUE #878: [Paths/module] module isn't working properly for Windows OS
Initially we are think that were an error from [Collections/split], as related on #872

More detailed what is happening here. (The problem isn't on this library itself), look:

I created a file called `paths.art` with this code:
image

Look what happens when I run it:

image

So, look at lines [1750](https://github.com/arturo-lang/arturo/blob/1136d74ff6685ec5b16c64021b3cf1a0dfc45fa8/src/library/Collections.nim#L1750) and [1797](https://github.com/arturo-lang/arturo/blob/1136d74ff6685ec5b16c64021b3cf1a0dfc45fa8/src/library/Collections.nim#L1797). 
Arturo is using the function `item.split(DirSep))`, `split` is a nim's standard function from `strutils.nim` module, none problem here, and `DirSep` that also is a nim's standard function comes from `osseps.nim` module, none problem here too.

But, let's see what `DirSep` returns for windows systems: 
```
  DirSep* =
    when defined(macos): ':'
    elif doslikeFileSystem or defined(vxworks): '\\'
    elif defined(RISCOS): '.'
    else: '/'
    ## The character used by the operating system to separate pathname
    ## components, for example: `'/'` for POSIX, `':'` for the classic
    ## Macintosh, and `'\\'` on Windows.
```
Yes, it'll split using a `\`. As Arturo's `module` function is returning with bot separators, it won't work. It'll separate just until user's path, and then a big string.

It's a issue related with [`module`](https://github.com/arturo-lang/arturo/blob/master/src/library/Paths.nim#L246) function.

_Originally posted by @RickBarretto in https://github.com/arturo-lang/arturo/issues/872#issuecomment-1353092615_

### PR DESCRIPTION
**Pull request by RickBarretto**
# Description

Adds support for windows path style.
For windows, both `\\` and `/` will be processed.

Fixes #878

On my local build:
image

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Add/Update unit tests

## Requires
- [ ] Build update
- [x] Unit test output update

### CODE REVIEW

**Comment by RickBarretto** [comment]
:rocket:

File: src/library/Collections.nim
Comment (drkameleon) [inline src/library/Collections.nim:1766]:
```suggestion
                        SetInPlace(newStringBlock(InPlaced.s.split({DirSep, AltSep})))
```

As I have just discovered myself, this is already supported by Nim: https://nim-lang.org/docs/os.html#AltSep

To keep things quite Nim-compliant, I think this could be solved even more easily. 😉

Comment (drkameleon) [reply src/library/Collections.nim:1766]: Since this is just for the literal, in-place use case of `split.path`, if you could add it as well for the "normal" version (below), it would be great! :)

**Review by drkameleon — CHANGES_REQUESTED** [review]

File: src/library/Collections.nim
Comment (RickBarretto) [reply src/library/Collections.nim]: Done!

Comment (drkameleon) [reply src/library/Collections.nim]: Awesome! :)

**Comment by drkameleon** [comment]
Is `split.path` working now for non-literal values as well?

P.S. sorry, I may be a bit slow to respond, but I'm taking some time offline until Monday morning, as usual

**Comment by RickBarretto** [comment]
What exactly means "non-literal values"?

**Comment by drkameleon** [comment]
> What exactly means "non-literal values" in this case?

I'm a bit confused: https://github.com/arturo-lang/arturo/blob/0fe08ea40f1a0ad3f18af2a919991755dc8cbc86/src/library/Collections.nim#L1809

This is the piece of code that would be called when doing something like `split.path "some/path"`... But I don't see any change...

**Comment by RickBarretto** [comment]
I got it what you're talking about.
image

**Comment by RickBarretto** [comment]
I'll fix it!

**Comment by RickBarretto** [comment]
Added some changes, I'll push now and cover it on unit-tests!
image

**Comment by RickBarretto** [comment]
> P.S. sorry, I may be a bit slow to respond, but I'm taking some time offline until Monday morning, as usual

No problem! 🙂

**Comment by RickBarretto** [comment]
Also, I'm seeing now that you removed the `when defined(windows)`, so this function is OS agnostic now.
So, I'll update the unit-tests.

**Comment by RickBarretto** [comment]
@drkameleon. What is breaking now. 
`DirSep`  `AltSep` are OS dependent, so, when I'm on windows, `DirSep` is `\\`, while `AltSep` is `/`, but, on Unix, `DirSep` is `/` and `AltSep` doesn't exists.

So, the better approach would be use `\\` or `/`, since `DirSep` and `AltSep` are just chars, anyway.

I'll wait for your reply about it.

**Comment by drkameleon** [comment]
> @drkameleon. What is breaking now: `DirSep` `AltSep` are OS dependent, so, when I'm on windows, `DirSep` is `\\`, while `AltSep` is `/`, but, on Unix, `DirSep` is `/` and `AltSep` doesn't exists.
> 
> So, the better approach would be use `\\` or `/`, since `DirSep` and `AltSep` are just chars, anyway.
> 
> I'll wait for your reply about it. But, imho, an OS agnostic `split.path` would be better.

Hmm... weird. I'll have to look into it.

Regarding `AltSep`: https://github.com/nim-lang/Nim/blob/version-1-6/lib/pure/includes/osseps.nim#L36-L38 (so, it should be just the same, only `{DirSep,AltSep}` being a set would get reduced to whatever it has to be, during compile-time)

**Comment by RickBarretto** [comment]
Alright, have a look and give me the feedback.

**Comment by drkameleon** [comment]
I think there is no problem at all with the implementation. The main problem is how we'd go about testing it and verifying its output.

In a few words: on Windows, `split.path "some\path"` returns `[some, path]` while on Linux/macOS, it returns the whole string (as `\` is not a path separator. On the other hand, on Linux/macOS, `split.path "some/path"` returns `[some, path]` while on Windows it would return the whole string. Given what is considered the "normal" path separator for each of the OSes, I don't think this is an undesired behavior.

Now, that being said, I think we just test it appropriately.

Let me see what I can come up with...

File: tests/unittests/lib.collections.art
Comment (drkameleon) [inline tests/unittests/lib.collections.art:712]:
```suggestion
    (sys\os == "windows")? [
        debug split.path "directory\\wofilerld"
        debug split.path "\\usr\\bin"

        p1: "directory\\wofilerld"
        p2: "\\usr\\bin"
    ][
        debug split.path "directory/wofilerld"
        debug split.path "/usr/bin"

        p1: "directory/wofilerld"
        p2: "/usr/bin"
    ]
```

Basically, we could use different paths based on the active OS (and I have removed the ones that wouldn't look very normal for Linux/macOS so that - if we insist on testing common output - the output can possibly be the same on all OSes...)

File: tests/unittests/lib.collections.art
Comment (drkameleon) [inline tests/unittests/lib.collections.art:719]:
```suggestion
    debug split.path p1
    debug split.path p2
    split.path 'p1, debug p1
    split.path 'p2, debug p2
```

Since I removed the `p3` from above, let's remove this one too...

File: tests/unittests/lib.collections.res
Comment (drkameleon) [inline tests/unittests/lib.collections.res:321]:
```suggestion
[directory wofilerld] :block 
[ usr bin] :block 
[directory wofilerld] :block 
[ usr bin] :block 
[directory wofilerld] :block 
[ usr bin] :block 
```

And - hopefully - this would be the expected output... (I so hope I haven't messed up... lol)

**Comment by drkameleon** [comment]
And another tiny issue I noticed (I had no idea there would be so many different issues with the implementation of such a seemingly ultra-simple functionality):

When we're splitting a path like "/usr/bin" the resulting array is: `["" "usr" "bin"]`, which is obviously *not* what we'd expect. So, I guess... we should also filter empty strings from the results. What do you think?

File: tests/unittests/lib.collections.art
Comment (RickBarretto) [reply tests/unittests/lib.collections.art:712]: Yes, I agree. But, since Windows uses the alt `/` too, I think it should be covered.

**Comment by RickBarretto** [comment]
> What do you think?

Anyway, we must to process the first item. Here or in another function.
1. Here, is just check if starts with `/` or `\\`. 
2. For another function check if the first item is a "", and then just ignore it.

But, I think that check it here is the best approach.

**Review by drkameleon — APPROVED** [review]

**Comment by RickBarretto** [comment]
> So, something like this, maybe (and its equivalent for InPlace values) ?

Like it, I will see a better approach since delete the first item of any array is a bad idea for performance.
Maybe, just ignore the first element?
The problem is that idk what Nim would do at the background using string split-range.

**Comment by drkameleon** [comment]
> > So, something like this, maybe (and its equivalent for InPlace values) ?
> 
> Like this, I will see a better approach since delete the first item of any array is a bad idea for performance. Maybe, just ignore the first element? The problem is that idk what Nim would do at the background using string split-range.

The truth is the *ideal* solution would be to have our own InPlace variants (from a Nim's perspective) for every possible InPlace Arturo function. In a few words, every `InPlaced.a = someArray` is fundamentally "wrong" (since it's not, properly-speaking, InPlace-modified). But I guess, it's something we cannot iron out along the way (I guess there are many such instances throughout the library).

In this case, I'll leave it up to you (provided that it behaves as it should obviously lol) 😄

**Comment by drkameleon** [comment]
That being said, after rethinking about it, given that the initial argument in this specific example (`split.path`) is a String (even though it may be passed as a Literal, indirectly), the result will always be a Block.

So... it can't possibly be modified InPlace. We'll have to create a new Block value. So... in that sense, it would be as efficient as not-using Literal values. I'm talking about the performance of doing something like: `split.path somePath` vs `split.path 'somePath`

**Comment by RickBarretto** [comment]
> So... it can't possibly be modified InPlace.

It's 100% true for this case.

**Comment by RickBarretto** [comment]
I'm taking a long time because I'm testing locally before push.

**Comment by RickBarretto** [comment]
I'm current trying with 
```nim
split(InPlaced.s[2 .. ^1], {DirSep, AltSep})
``` 
But it's not working as I expected.

**Comment by drkameleon** [comment]
> ```nim
> split(InPlaced.s[1 .. ^1], {DirSep, AltSep})
> ```

if `s[0]` is indeed confirmed to be a `{DirSep,AltSep}`, I honestly don't see how it's not working... What are we missing?! 🤔

**Comment by drkameleon** [comment]
Just an idea...

Could you try it like:

```nim
let str = InPlaced.s
split(str[1..^1], {DirSep,AltSep})
```

at least, this way we could confirm whether it has to do with the internal implementation of this `InPlaced`...

**Comment by RickBarretto** [comment]
> Could you try it like:

Alright, I'll try it!

I tried it on Nim's Playgroud:
  
image

**Comment by RickBarretto** [comment]
I'm, I got what I was done wrong. I'll fix it!

**Comment by RickBarretto** [comment]
Alright, fixed! 
image

But, I'll remove the last behavior, make it work as the first. What do you think?

**Comment by RickBarretto** [comment]
Alright, fixed too! I'll take a print, send here and finally push! :rocket:

**Comment by RickBarretto** [comment]
On my local build:
image

**Comment by RickBarretto** [comment]
image
🚀

**Comment by drkameleon** [comment]
> I got what I was done wrong. I'll fix it!

Saw the code (the fix) and totally got what was going on... Damn! Good job with that! 😉

**Review by drkameleon — APPROVED** [review]
Pure awesomeness! Very good job again on this one! 😄 

Ready to merge! 🚀

**Comment by RickBarretto** [comment]
Thank you!

**Pull request merged**
